lib/exmeralda/topics/generate_embeddings_worker.ex

```diff
@@ -2,99 +2,41 @@ defmodule Exmeralda.Topics.GenerateEmbeddingsWorker do
   use Oban.Worker, queue: :ingest, max_attempts: 20
 
   alias Exmeralda.Repo
-  alias Exmeralda.Topics
   alias Exmeralda.Topics.{Chunk, Ingestion, Rag}
+  alias Exmeralda.Topics
 
   import Ecto.Query
 
-  @embeddings_batch_size 20
-
-  @impl Oban.Worker
-  def perform(%Oban.Job{
-        id: id,
-        args: %{"library_id" => library_id, "ingestion_id" => ingestion_id}
-      }) do
-    Repo.transact(fn ->
-      with {:ok, ingestion} <- fetch_ingestion(ingestion_id, library_id: library_id) do
-        {:ok,
-         from(c in Chunk, where: c.ingestion_id == ^ingestion_id, select: c.id)
-         |> Repo.all()
-         |> Enum.chunk_every(@embeddings_batch_size)
-         |> Enum.map(
-           # Passing the parent job_id so it's easier to find which children chunk jobs
-           # were enqueued by this worker.
-           &__MODULE__.new(%{chunk_ids: &1, ingestion_id: ingestion.id, parent_job_id: id})
-         )
-         |> Oban.insert_all()}
-      end
-    end)
-    |> case do
-      {:error, :ingestion_not_found} ->
-        {:cancel, :ingestion_not_found}
-
-      {:error, {:ingestion_in_invalid_state, state}} ->
-        {:cancel, {:ingestion_in_invalid_state, state}}
-
-      {:ok, _} ->
-        :ok
-    end
-  end
-
   def perform(%Oban.Job{args: %{"chunk_ids" => ids, "ingestion_id" => ingestion_id}} = job) do
-    Repo.transact(fn ->
-      with {:ok, ingestion} <- fetch_ingestion(ingestion_id),
-           {:ok, embeddings} <- generate_embeddings(ingestion, ids) do
-        embeddings
-        |> Enum.map(&Chunk.set_embedding(Map.put(&1, :embedding, nil), &1.embedding))
-        |> Enum.each(&Repo.update!/1)
-
-        if all_chunks_embedded?(ingestion.id) do
-          ingestion = Topics.update_ingestion_state!(ingestion, :ready)
-          {:ok, _} = Topics.mark_ingestion_as_active(ingestion.id)
-        end
-
-        {:ok, ingestion}
+    with {:ok, ingestion} <- fetch_ingestion(ingestion_id),
+         {:ok, embeddings} <- generate_embeddings(ids, ingestion, job) do
+      embeddings
+      |> Enum.map(&Chunk.set_embedding(Map.put(&1, :embedding, nil), &1.embedding))
+      |> Enum.each(&Repo.update!/1)
```

Comment:
If line 15, the `Repo.update!`, fails for any chunk, right now, we would not roll back. This means we would have the embedding set on some chunks and on some not.
Would it not be nicer, to roll back in that case?

(I guess errors are not likely to happen there, but the DB could be down or so)

The check if all chunks are embedded could then still be outside of the transaction.

WDYT?

Reply:
hmm. I just realised, when one of the jobs of the generate_embeddings_worker fails, we would also end up with partially embedded chunks.

How do we recover from such a state? Is there a way to retrigger the embedding for all unembedded chunks only?

Reply:
If the `Repo.update!` raises, or the generally if the worker errors, the worker would retry. So we should recover from errors that are not systematic (like a database down). If the worker fails for 20 times in a row then something is wrong and we probably need to fix a bug somewhere. In this case one could retry manually the workers in the oban admin. 

There is also this ticket https://bitcrowd.atlassian.net/browse/EX-117 which should allow admins to delete the ingestion in this case (at the moment you cannot delete an ingestion in `embedding` state).

In any case: I think we can filter out chunks that are already embedded in the `generate_embeddings` function so that if the worker retries, it only tries to embed the missing chunks 🪄